[Notification] CRUD notification APIs

qnkhuat opened the pull request:
Add 2 APIs:
- POST /api/notification
- PUT /api/notification/:id
  - Can also update/insert/delete all its sub-resources: card_notification, subscriptions, handlers, handler recipients


Examples


**Create an alert:**

- send once if above goal
- has 1 cron schedule
- has 2 handlers that use default template
    - to webhook channel
    - to email channel
        - send to one internal user, one external user

Make a POST request  to `/api/notification`
```json
{
	"payload_type": "notification/card",
	"active": true,
	"creator_id": 3,
	"payload": {
		"card_id": 2,
		"send_condition": "goal_above",
		"send_once": true
	},
	"subscriptions": [{
		"type": "notification-subscription/cron",
		"cron_schedule": "0 0 0 * * ?"
	}],
	"handlers": [{
			"channel_type": "channel/http",
			"channel_id": 1
		},
		{
			"channel_type": "channel/email",
			"recipients": [{
					"type": "notification-recipient/user",
					"user_id": 3
				},
				{
					"type": "notification-recipient/raw-value",
					"details": {
						"value": "ngoc@metabase.com"
					}
				}
			]
		}
	]
}
```

This endpoint will return a fully hydrated notification.

```json
{
  "payload_id": 72,
  "payload": {
    "id": 72,
    "card_id": 2,
    "send_once": true,
    "send_condition": "goal_above"
  },
  "creator": {
    "email": "crowberto@metabase.com",
    "first_name": "Crowberto",
    "last_login": null,
    "is_qbnewb": true,
    "is_superuser": true,
    "id": 3,
    "last_name": "Corv",
    "date_joined": "2024-12-05T04:57:26.47935Z",
    "common_name": "Crowberto Corv"
  },
  "payload_type": "notification/card",
  "handlers": [
    {
      "template_id": null,
      "channel_type": "channel/metabase-test",
      "channel": {
        "id": 3,
        "name": "Test channel",
        "description": "Test channel description",
        "type": "channel/http",
        "active": true
      },
      "template": null,
      "recipients": [
        {
          "type": "notification-recipient/user",
          "permissions_group_id": null,
          "details": null,
          "id": 82,
          "user_id": 3,
          "notification_handler_id": 129,
          "user": {
            "email": "crowberto@metabase.com",
            "first_name": "Crowberto",
            "last_login": null,
            "is_qbnewb": true,
            "is_superuser": true,
            "id": 3,
            "last_name": "Corv",
            "date_joined": "2024-12-05T04:57:26.47935Z",
            "common_name": "Crowberto Corv"
          }
        }
      ],
      "channel_id": 3,
      "notification_id": 72,
      "active": true,
      "id": 129
    },
    {
      "template_id": null,
      "channel_type": "channel/email",
      "channel": null,
      "template": null,
      "recipients": [
        {
          "type": "notification-recipient/user",
          "permissions_group_id": null,
          "details": null,
          "id": 83,
          "user_id": 3,
          "notification_handler_id": 130,
          "user": {
            "email": "crowberto@metabase.com",
            "first_name": "Crowberto",
            "last_login": null,
            "is_qbnewb": true,
            "is_superuser": true,
            "id": 3,
            "last_name": "Corv",
            "date_joined": "2024-12-05T04:57:26.47935Z",
            "common_name": "Crowberto Corv"
          }
        }
      ],
      "channel_id": null,
      "notification_id": 72,
      "active": true,
      "id": 130
    }
  ],
  "creator_id": 3,
  "subscriptions": [
    {
      "id": 125,
      "notification_id": 72,
      "type": "notification-subscription/cron",
      "event_name": null,
      "cron_schedule": "0 0 0 * * ?"
    }
  ],
  "active": true,
  "id": 72
}

```

**Update alert**
- delete the HTTP handler
- add a slack handler that sends to #general
```diff
{
  "payload_id": 72,
  "payload": {
    "id": 72,
    "card_id": 2,
    "send_once": true,
    "send_condition": "goal_above"
  },
  "creator": {
    "email": "crowberto@metabase.com",
    "first_name": "Crowberto",
    "last_login": null,
    "is_qbnewb": true,
    "is_superuser": true,
    "id": 3,
    "last_name": "Corv",
    "date_joined": "2024-12-05T04:57:26.47935Z",
    "common_name": "Crowberto Corv"
  },
  "payload_type": "notification/card",
  "handlers": [
-    {
-      "template_id": null,
-      "channel_type": "channel/http",
-      "channel": {
-        "id": 3,
-        "name": "Test channel",
-        "description": "Test channel description",
-        "type": "channel/http",
-        "active": true
-      },
-      "template": null,
-      "recipients": [
-        {
-          "type": "notification-recipient/user",
-          "permissions_group_id": null,
-          "details": null,
-          "id": 82,
-          "user_id": 3,
-          "notification_handler_id": 129,
-          "user": {
-            "email": "crowberto@metabase.com",
-            "first_name": "Crowberto",
-            "last_login": null,
-            "is_qbnewb": true,
-            "is_superuser": true,
-            "id": 3,
-            "last_name": "Corv",
-            "date_joined": "2024-12-05T04:57:26.47935Z",
-            "common_name": "Crowberto Corv"
-          }
-        }
-      ],
-      "channel_id": 3,
-      "notification_id": 72,
-      "active": true,
-      "id": 129
-    },
+    {
+      "id": -1,
+		"channel_type": "channel/slack",
+		"recipients": [{
+			"type": "notification-recipient/raw-value",
+			"details": {"value": "#general"}
+		}]
+	 },
    {
      "template_id": null,
      "channel_type": "channel/email",
      "channel": null,
      "template": null,
      "recipients": [
        {
          "type": "notification-recipient/user",
          "permissions_group_id": null,
          "details": null,
          "id": 83,
          "user_id": 3,
          "notification_handler_id": 130,
          "user": {
            "email": "crowberto@metabase.com",
            "first_name": "Crowberto",
            "last_login": null,
            "is_qbnewb": true,
            "is_superuser": true,
            "id": 3,
            "last_name": "Corv",
            "date_joined": "2024-12-05T04:57:26.47935Z",
            "common_name": "Crowberto Corv"
          }
        }
      ],
      "channel_id": null,
      "notification_id": 72,
      "active": true,
      "id": 130
    }
  ],
  "creator_id": 3,
  "subscriptions": [
    {
      "id": 125,
      "notification_id": 72,
      "type": "notification-subscription/cron",
      "event_name": null,
      "cron_schedule": "0 0 0 * * ?"
    }
  ],
  "active": true,
  "id": 72
}


```

qnkhuat committed:
- notification/alert -> notification/card
- notification/dashboard-subscription -> notification/dashboard

qnkhuat committed:
cljfmt

qnkhuat committed:
fix file not found

qnkhuat committed:
get a notification api and tests

qnkhuat committed:
test fixes

qnkhuat committed:
cljfmt

qnkhuat committed:
Merge branch 'a_brave_new_alert' into notification-apis

qnkhuat committed:
copy and serdes test fixes

qnkhuat committed:
serdes test

qnkhuat committed:
TBC

qnkhuat committed:
Merge branch 'a_brave_new_alert' into notification-apis

qnkhuat committed:
Merge branch 'notification-apis' into notification-apis-2

qnkhuat committed:
add notification API

qnkhuat committed:
run_once => send_once

qnkhuat committed:
Merge branch 'notification-apis' into notification-apis-2

qnkhuat committed:
Apply suggestions from code review

Co-authored-by: Chris Truter <crisptrutski@users.noreply.github.com>

qnkhuat committed:
Apply suggestions from code review

Co-authored-by: Chris Truter <crisptrutski@users.noreply.github.com>

qnkhuat committed:
Chris's suggestions

qnkhuat committed:
Chris's suggestions

qnkhuat committed:
spec update and PUT API

qnkhuat commented on src/metabase/models/spec_update.clj:
I'll write tests for this but I'd love to see people's feeling about this first.

I originally write the update for notification manually
- walk the payload, check changes, update it
- walk the handlers, check changes, update it

The I realized the code is too big and hard to manage, so I came up with this, the idea here is you define a dependency spec, then the code will walk and update all its sub-resources for you. see [this](https://github.com/metabase/metabase/pull/50921/files#diff-076a8437ddc0810830d668cc7d7e703bd6527a2dbc3a5a2605d728b1b4bec3fcR432-R451) for an example of the spec.

qnkhuat replied:
Also, the logging is quite nice, an example


```
2024-12-06 07:54:14,990 DEBUG models.spec-update :: root: enter
2024-12-06 07:54:14,990 DEBUG models.spec-update :: root: single row spec found
2024-12-06 07:54:14,990 DEBUG models.spec-update :: root: no change detected for :model/Notification 2
2024-12-06 07:54:14,990 DEBUG models.spec-update :: root: nested models detected, updating nested models :model/Notification 2
2024-12-06 07:54:14,990 DEBUG models.spec-update :: root / 2: enter
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload: enter
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload / root: enter
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload / root: single row spec found
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload / root: no change detected for :model/NotificationCard 2
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload / root: exit
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / payload: exit
2024-12-06 07:54:14,991 DEBUG models.spec-update :: root / 2 / subscriptions: enter
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / subscriptions / root: enter
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / subscriptions / root: multi-row spec found
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / subscriptions / root: exit
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / subscriptions: exit
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / handlers: enter
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / handlers / root: enter
2024-12-06 07:54:14,992 DEBUG models.spec-update :: root / 2 / handlers / root: multi-row spec found
2024-12-06 07:54:14,993 DEBUG models.spec-update :: root / 2 / handlers / root: nested spec found, creating rows one by one
2024-12-06 07:54:14,995 DEBUG models.spec-update :: root / 2 / handlers / root: Created a new entity :model/NotificationHandler 4
2024-12-06 07:54:14,996 DEBUG models.spec-update :: root / 2 / handlers / root / 4: enter
2024-12-06 07:54:14,997 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients: enter
2024-12-06 07:54:14,997 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients / root: enter
2024-12-06 07:54:14,997 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients / root: multi-row spec found
2024-12-06 07:54:14,997 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients / root: no nested spec found, batch creating 1 new rows of :model/NotificationRecipient
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients / root: exit
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root / 4 / recipients: exit
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root / 4: exit
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root: Attempt updating 1 rows of :model/NotificationHandler
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root / 3: enter
2024-12-06 07:54:14,998 DEBUG models.spec-update :: root / 2 / handlers / root / 3: Updating {:notification_id 2, :channel_type "channel/email", :channel_id nil, :template_id 2, :active true}
2024-12-06 07:54:15,000 DEBUG models.spec-update :: root / 2 / handlers / root / 3: nested models detected, updating
2024-12-06 07:54:15,000 DEBUG models.spec-update :: root / 2 / handlers / root / 3 / recipients: enter
2024-12-06 07:54:15,000 DEBUG models.spec-update :: root / 2 / handlers / root / 3 / recipients / root: enter
2024-12-06 07:54:15,000 DEBUG models.spec-update :: root / 2 / handlers / root / 3 / recipients / root: multi-row spec found
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2 / handlers / root / 3 / recipients / root: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2 / handlers / root / 3 / recipients: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2 / handlers / root / 3: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2 / handlers / root: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2 / handlers: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root / 2: exit
2024-12-06 07:54:15,001 DEBUG models.spec-update :: root: exit
```

qnkhuat committed:
enter root at, well root

qnkhuat committed:
Merge branch 'notification-apis' into notification-apis-2

qnkhuat committed:
unseed

qnkhuat committed:
add tests

qnkhuat committed:
add error tests

qnkhuat committed:
delete silly logs

qnkhuat committed:
test fixes

qnkhuat committed:
stable handler and recipients by type

qnkhuat committed:
use find first

qnkhuat committed:
sort by type

qnkhuat commented on src/metabase/models/notification.clj:
```diff
@@ -391,3 +428,29 @@
+                                                              :compare-row #(select-keys % [:notification_handler_id :type :user_id :permissions_group_id :details])
+                                                              :multi-row?  true}}}}})
+
+(defn update-notification!
```
the 2nd model is unnatural because system event doesn't have an equivalent table

crisptrutski replied:
Can you refresh my memory on why NotificationCard needs to be inserted first? It sounds like a M2M join table, so why is the relationship inverted?

It feels a bit strange for the spec nesting to be different to the database structure.

Maybe you could just define both as top level specifications, with a keyword instead of a map for references to non "private" specs?

```
- notification_card
  - *notification*
- notification
  - handlers
  - subscriptions
```

piranha replied:
@qnkhuat so the fact that `NotificationSubscription` also needs to be created before `Notification` is fine, but for `NotificationCard` it's a problem?

qnkhuat replied:
maybe I missed word somewhere but `NotificationSubscription` doesn't need to be created before, in fact it has to be created after.

Why does `notification_card` need to be created first? It is because of `notification.payload_id` implicitly is a FK to `notification_card.id`

crisptrutski replied:
Ah yeah I agree this is not a good enough reason to flip the more natural hierarchy.

Post-processing might be the cleanest solution here, as ugly as that is.

crisptrutski replied:
What about using an entity ID for the relationship instead?

qnkhuat replied:
Could you guys clarify what do you mean by post-processing? if any I think we need preprocessing right? in this case we'need to pre-insert notification-card, so that we can use spec to update properly

crisptrutski replied:
I mean inserting the notification first, without the payload id, then updating the id afterwards. A bit awkward with the defensive hooks, but I think this is more intuitive and generalizable than reshuffling the creation order.

qnkhuat replied:
ah okay. but that's problem for another PR where we replace `create-nofication!` using this spec. are there anything we need to resolve before moving forward here?

crisptrutski replied:
nothing blocking. you started this thread, feel free to resolve it 😄

qnkhuat committed:
requires :id

crisptrutski commented on src/metabase/models/spec_update.clj:
```diff
@@ -0,0 +1,172 @@
+(ns metabase.models.spec-update
```
A bit of a quibble, but what do you think about moving this to a `metabase.models.util.spec-update` namespace? I just worry that "top level" model namespaces look too much like actual models.

qnkhuat committed:
Merge branch 'a_brave_new_alert' into notification-apis-2

crisptrutski commented on src/metabase/models/notification.clj:
```diff
@@ -391,3 +428,29 @@
+   ;; a function that takes a row and returns a map of the columns to compare
+   :compare-row  #(select-keys % [:active])
+   :nested-specs {:payload       {:model        :model/NotificationCard
+                                  :compare-row  #(select-keys % [:send_condition :send_once])}
```
I'd much rather the spec itself was pure-data.

```suggestion
                                  :compare-values [:send_condition :send_once]}
```

piranha replied:
+1 here, at least until there is an obvious need for custom processing

Also maybe not `:compare-values`, but something like `:row-id` or `:unique-id`?

qnkhuat replied:
how about :compare-cols?

crisptrutski commented on src/metabase/models/spec_update.clj:
```diff
@@ -0,0 +1,172 @@
+(defmacro ^:private with-enter-path
+  [path & body]
+  `(binding [*current-path* (conj *current-path* ~path)]
+     (log/debugf "%s enter" (current-path))
```
These should be `tracef` IMHO

crisptrutski commented on src/metabase/models/spec_update.clj:
```diff
@@ -0,0 +1,172 @@
+
+(defmacro ^:private with-enter-path
+  [path & body]
+  `(binding [*current-path* (conj *current-path* ~path)]
```
It doesn't seem like the code would be that much trickier if it just worked with a path variable. I'd much prefer not using dynamic state.

piranha replied:
Huge +1, just make this an explicit argument and it'll be easier to debug too.

crisptrutski commented on src/metabase/models/spec_update.clj:
```diff
@@ -0,0 +1,172 @@
+    (when (seq to-update)
+      (log/debugf "%s Attempt updating %d rows of %s" (current-path) (count to-update) model)
+      (doseq [row to-update]
+        (with-enter-path (:id row)
+          (let [new-row (sanitize-row row)]
+            (log/debugf "%s Updating" (current-path))
+            (t2/update! model (:id row) new-row))
```
Just as an example

```suggestion
        (let [path    (conj path (:id row))
              new-row (sanitize-row row)]
            (log/debugf "%s Updating" path)
            (t2/update! model (:id row) new-row))
```

qnkhuat committed:
Merge branch 'a_brave_new_alert' into notification-apis-2

qnkhuat committed:
Merge branch 'a_brave_new_alert' into notification-apis-2

crisptrutski reviewed (commented):
On one hand I'm a little uncomfortable extracting such a generalized mechanism based on one case we've had so far, but on the other I like taking a data driven approach when there are so many details to get right.

How realistically do you think we'll have other use cases with these same semantics? Are there existing complex model relationships that would benefit from refactoring to this system?

piranha commented:
@crisptrutski TBH what we see here seems like a great first step into "automated CRUD endpoints" we were talking at some point, aside from `NotificationCard` issues. :)

qnkhuat commented:
> Are there existing complex model relationships that would benefit from refactoring to this system?

to the best of my knowledge we don't have this level of complex models, the most complicated I can think of is `database -> table -> field`, but they are managed by sync.

> On one hand I'm a little uncomfortable extracting such a generalized mechanism based on one case we've had so far,
I'm happy to take this local to notification, but I think we'll attract more usage by putting it out there, which will improve this further.

qnkhuat committed:
- Move spec-update to models/util/spec_update
- compare-row is a list, not a fn
- no more dynamic current path, use explit path

piranha commented on src/metabase/api/notification.clj:
```diff
@@ -9,12 +9,36 @@
+  {body models.notification/FullyHydratedNotification}
+  (models.notification/hydrate-notification
+   (models.notification/create-notification!
+    (dissoc body :handlers :subscriptions)
```
the fact that `FullyHydratedNotification` is not closed (and maybe it should be) gives me some feeling of uneasiness just passing whatever here. :) TBH I'd rather see `select-keys` than `dissoc` in places like that, or _at least_ a closed map.

qnkhuat replied:
yea I dislike it so much, but last I checked in malli if you have a closed map, and it's merge with another map or dispatch it no longer closed.

I plan to revisit this and see if we can have a workaround, also I plan to use convert to use registry so we don't have to re-eval all methods when we change one schema.

piranha commented on src/metabase/models/notification.clj:
```diff
@@ -89,14 +90,15 @@
   [:merge
    [:map
-    [:payload_type (into [:enum] notification-types)]]
+    [:payload_type (apply ms/enum-decode-keyword notification-types)]]
```
Why `apply` everywhere if it's your function, maybe it should just accept a list? :)

piranha commented on src/metabase/models/util/spec_update.clj:
```diff
@@ -0,0 +1,162 @@
+    (when (seq to-create)
+      (if nested-specs
+        (do
+          (log/debugf "%s nested spec found, creating rows one by one" (format-path path))
```
I agree with Chris, this should be `tracef`

qnkhuat replied:
done

crisptrutski reviewed (approved):
Agree with Sanya's remaining points, but nothing blocking for me

sansolovyov reviewed (approved):
I like that!

qnkhuat committed:
sanya's feedback

qnkhuat commented on src/metabase/models/util/spec_update.clj:
```diff
@@ -0,0 +1,162 @@
+(defn- with-parent-id
+  [row nested-specs parent-id]
+  (merge row
+         (into {}
```
right!
